src/stim/search/sat/wcnf.cc

```diff
@@ -0,0 +1,261 @@
+    }
+    if (x.variable == BOOL_LITERAL_TRUE) {
+      return ~y;
+    }
+    if (y.variable == BOOL_LITERAL_FALSE) {
+      return x;
+    }
+    if (y.variable == BOOL_LITERAL_TRUE) {
+      return ~x;
+    }
+    BoolRef z = new_bool();
+    // Forbid strings (x, y, z) such that z != XOR(x, y)
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (0, 0, 1)
+      clause.add_var(x);
+      clause.add_var(y);
+      clause.add_var(~z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (0, 1, 0)
+      clause.add_var(x);
+      clause.add_var(~y);
+      clause.add_var(z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (1, 0, 0)
+      clause.add_var(~x);
+      clause.add_var(y);
+      clause.add_var(z);
+      add_clause(clause);
+    }
+    {
+      Clause clause;
+      // hard clause (x, y, z) != (1, 1, 1)
+      clause.add_var(~x);
+      clause.add_var(~y);
+      clause.add_var(~z);
+      add_clause(clause);
+    }
+    return z;
+  }
+
+  size_t quantized_weight(bool weighted, size_t weight_scale_factor, size_t top, Weight weight) {
+    if (weight == HARD_CLAUSE_WEIGHT) {
+      // Hard clause
+      return top;
+    }
+    // Soft clause
+    if (!weighted) {
+      // For unweighted problems, all soft clauses have weight 1.
+      return 1;
+    }
+    return std::round(weight / max_weight * (double)weight_scale_factor);
+  }
+
+  std::string to_wdimacs(bool weighted, size_t weight_scale_factor) {
+    // 'top' is a special weight used to indicate a hard clause.
+    // Should be at least the sum of the weights of all soft clauses plus 1.
+    size_t top;
+    if (weighted) {
+      top = 1 + weight_scale_factor * clauses.size();
+    } else {
+      top = 1 + clauses.size();
+    }
+
+    // WDIMACS header format: p wcnf nbvar nbclauses top
+    // see http://www.maxhs.org/docs/wdimacs.html
+    std::stringstream ss;
+    ss << "p wcnf "<< num_variables << " "<<clauses.size() << " " << top << "\n";
+
+    // Add clauses, 1 on each line.
+    for (const auto& clause : clauses) {
+      // WDIMACS clause format: weight var1 var2 ...
+      // To show negation of a variable, the index should be negated.
+      size_t qw = quantized_weight(weighted, weight_scale_factor, top, clause.weight);
+      // There is no need to add a clause with zero weight.
+      // This can happen if the error has probability 0.5 or if the weight_scale_factor is too
+      // small to accomodate the entire dynamic range of weight values.
+      if (qw == 0) continue;
+      ss << qw;
+      for (size_t i=0; i<clause.vars.size(); ++i) {
+        BoolRef var = clause.vars[i];
+        // Variables are 1-indexed
+        if (var.negated) {
+          ss << " -" << (var.variable + 1);
+        } else {
+          ss << " " << (var.variable + 1);
+        }
+      }
+      // Each clause ends with 0
+      ss << " 0\n";
+    }
+    return ss.str();
```

Comment:
Nit: I'd refactor the method into a `write_dimacs` style, taking a `std::ostream &out` parameter, instead of forcing the allocation of the string. Not necessary.

Reply:
Ok, thanks.